tensorrt_llm/_torch/auto_deploy/transform/library/mxfp4_moe.py

```diff
@@ -0,0 +1,339 @@
+    next_states = (
+        next_states * routing_weights.transpose(0, 1).view(num_experts, batch_size, -1)[..., None]
+    )
+    next_states = next_states.sum(dim=0)  # [B, S, H]
+    return next_states
+
+
+def _moe_dense_mlp_repl(
+    hidden_states: torch.Tensor,
+    routing_weights: torch.Tensor,
+    gate_up_w: torch.Tensor,
+    gate_up_b: torch.Tensor,
+    down_w: torch.Tensor,
+    down_b: torch.Tensor,
+    alpha: float,
+    limit: float,
+    minus_limit: float,
+) -> torch.Tensor:
+    return torch.ops.auto_deploy.torch_moe_dense_mlp(
+        hidden_states, routing_weights, gate_up_w, gate_up_b, down_w, down_b, alpha, limit
+    )
+
+
+@TransformRegistry.register("match_dense_moe_pattern")
+class MatchMOEDenseMLP(BaseTransform):
+    def _apply(
+        self,
+        gm: GraphModule,
+        cm,
+        factory,
+        shared_config,
+    ) -> Tuple[GraphModule, TransformInfo]:
+        graph = gm.graph
+        patterns = ADPatternMatcherPass()
+
+        B, S, H = 2, 4, 8  # batch, seq, hidden
+        E, In = 3, 16  # experts, intermediate (I); gate_up has 2I
+        T = B * S
+
+        dummy_args = [
+            torch.randn(B, S, H, device="meta", dtype=torch.float16),  # hidden_states
+            torch.randn(T, E, device="meta", dtype=torch.float16),  # routing_weights
+            torch.randn(E, H, 2 * In, device="meta", dtype=torch.float16),  # gate_up_w  [E,H,2I]
+            torch.randn(E, 2 * In, device="meta", dtype=torch.float16),  # gate_up_b  [E,2I]
+            torch.randn(E, In, H, device="meta", dtype=torch.float16),  # down_w     [E,I,H]
+            torch.randn(E, H, device="meta", dtype=torch.float16),  # down_b     [E,H]
+            1.07,
+            10.1,
+            -10.1,
+        ]
+
+        op_ignore_types = {
+            torch.ops.aten.view.default: (int,),
+            torch.ops.aten.reshape.default: (int,),
+            torch.ops.aten.repeat.default: (int,),
+            torch.ops.aten.slice.Tensor: (int,),
+            torch.ops.aten.unsqueeze.default: (int,),
+            torch.ops.aten.transpose.int: (int,),
+        }
+
+        scalar_workaround = {"alpha": 1.07, "limit": 10.1, "minus_limit": -10.1}
+
+        register_ad_pattern(
+            search_fn=_moe_dense_mlp_pattern,
+            replace_fn=_moe_dense_mlp_repl,
+            patterns=patterns,
+            dummy_args=dummy_args,
+            op_ignore_types=op_ignore_types,
+            scalar_workaround=scalar_workaround,
+        )
+
+        num_matches = patterns.apply(graph)
+        info = TransformInfo(
+            skipped=False,
+            num_matches=num_matches,
+            is_clean=False,
+            has_valid_shapes=False,
+        )
+        return gm, info
+
+
+def _get_param(gm: GraphModule, name: str) -> torch.Tensor:
+    return gm.get_parameter(name)
+
+
+def _get_submodule_of_param(gm: GraphModule, param_name: str) -> Tuple[nn.Module, str, str]:
```

Comment:
nit: the utility should work for `nn.Module` more generally, not just `GraphModule`